### PR TITLE
"newmm-safe" option -- fix newmm issue, take too long time with some long text

### PR DESCRIPTION
**Pull request by bact**
### The problem
- The issue is discribed in #241
- For example, if an input text is "หน้าด้านหน้าด้านหน้าด้านหน้าด้านหน้าด้าน", as "หน้าด้าน" and "ด้านหน้า" are both valid word (according to the dictionary), lots of word graphs representing different possibility of word breaks will be created.
- If the ambiguos text is very long, it will take so long time to finished the job, as the possibility will grow explonetially.

### Proposed solution
- Preprocess the input text before the actual tokenization
- if the input text is longer than the limit (`TEXT_LIMIT`) with window size left and right), breaks them into smaller chunks (`text_parts`) then tokenizes each chunk
- when breaks the text, the breaking point should fall at valid word/syllable ends
  - if there's a space, use the right-most space as a breaking point
  - if there's no space, current implementation find candidate breaking points by feed a smaller sample text in a window (from `TEXT_LIMIT - TEXT_SCAN_LEFT` to `TEXT_LIMIT + TEXT_SCAN_RIGHT`) to newmm again - and choose to break after the longest token
- This preprocess can be enable/disable by `safe_mode` parameter.
  - If `safe_mode` is `True`, do the preprocess (will be slower)
  - If `safe_mode` is `False` (default), don't do the preprocess (may face long wait for edge cases)

### Effects from the proprosed solution
- chunk size, window size, and complexity of finding valid breaking strategy can affect the speed and accurary
- current chunk size is 100 +- 20
  - too large chunk size = still long text, will not solve to issue --> affect speed
  - too small chunk size = more work to do --> affect speed, and maybe accuracy as well
  - too large window size = more work to do --> affect speed
  - too small window size = can break a single word into two words --> affect accuray
- need to check the performance with the `pythainlp.benchmarks`

### Alternative solution
- The proposed preprocssing can be alternatively provided as a separated function that the user needs to call by themselves, if they want a safe tokenization.
- Another alternative, and more elegant, solution is maybe to limit nodes in word graph to a certian number. (Do it directly inside `_bfs_paths_graph()`)

### CODE REVIEW

File: pythainlp/tokenize/ssg.py [hunk 1]
```diff
@@ -5,4 +5,7 @@
 
 
 def segment(text: str) -> List[str]:
+    if not text or not isinstance(text, str):
```
Comment (heytitle) [inline pythainlp/tokenize/ssg.py]: This condition appears [many times][ref] in the code. I think we should consider refactor it.


[ref]: https://github.com/PyThaiNLP/pythainlp/search?q=%22if+not+text+or+not+isinstance%28text%2C+str%29%22&unscoped_q=%22if+not+text+or+not+isinstance%28text%2C+str%29%22

**Comment by bact** [comment]
Results from the code in this notebook
https://colab.research.google.com/drive/15BYn_XOouej4M3CDL6ltO1j5uAq99Whq

## Length of test texts (6 texts in total)
- 16
- 99
- 99
- 392
- 576
- 1848

## From dev branch
- 10000 loops, best of 3: 31.4 µs per loop
- 1000 loops, best of 3: 231 µs per loop
- 1000 loops, best of 3: 263 µs per loop
- 100 loops, best of 3: 2.58 ms per loop
- (Cannot finished)
- (Cannot finished)

## From fix-newmm-longtext
- 10000 loops, best of 3: 31.4 µs per loop
- 1000 loops, best of 3: 235 µs per loop
- 1000 loops, best of 3: 263 µs per loop
- 1000 loops, best of 3: 1.72 ms per loop
- 1 loop, best of 3: 285 ms per loop
- 1 loop, best of 3: 2.76 s per loop

File: pythainlp/tokenize/ssg.py (hunk 1 above)
Comment (wannaphongcom) [reply pythainlp/tokenize/ssg.py]: @heytitle `isinstance(text, str)` only  ?

**Comment by wannaphongcom** [comment]
## From PyThaiNLP 2.0.7 (use marisa-trie)

1. 10000 loops, best of 3: 26.9 µs per loop
2. 1000 loops, best of 3: 232 µs per loop
3. 1000 loops, best of 3: 248 µs per loop
4. 1000 loops, best of 3: 813 µs per loop
5. 1000 loops, best of 3: 1.04 ms per loop
6. 100 loops, best of 3: 2.56 ms per loop
7. (Cannot finished)
8. (Cannot finished)
9. 10 loops, best of 3: 90.2 ms per loop

Colab : https://colab.research.google.com/drive/19fy_hZtdhGpDY_hpwLO4lR2J8ajzuLY_

**Comment by wannaphongcom** [comment]
- [ ] check the performance with the pythainlp.benchmarks

**Comment by wannaphongcom** [comment]
## Tokenisation Speed Benchmark

Notebook by @heytitle 

- dev [Notebook](https://colab.research.google.com/drive/1nSKEFyqqddaScvC8ppmLeQVI0I002klF)
- fix-newmm-longtext [Notebook](https://colab.research.google.com/drive/1Zu5Fu40p7cNuaLgKZplZcFZ5nbuTCgm9)

**Comment by bact** [comment]
> ## Tokenisation Speed Benchmark

  | method | Char(200) | Char(500) | Char(1000) | Char(2000)
-- | -- | -- | -- | -- | --
0 | newmm (dev) | 0.0008±0.0000 | 0.0017±0.0002 | 0.0031±0.0001 | 0.0058±0.0004
1 | newmm (fix-newmm-longtext) | 0.0009±0.0001 | 0.0018±0.0000 | 0.0046±0.0015 | 0.0080±0.0016

File: pythainlp/tokenize/ssg.py (hunk 1 above)
Comment (heytitle) [reply pythainlp/tokenize/ssg.py]: I mean this `if`-condition appears many times.

**Comment by wannaphongcom** [comment]
## check the performance with the pythainlp.benchmarks

**dev**

Colab : [Google colab](https://colab.research.google.com/drive/1pNqNFbP7nd1lRPWDkbpuQI10zepXG8Eo)

```
============== Benchmark Result ==============
               metric      mean±std       min    max
        char_level:tp   25.49±27.32  1.000000  240.0
        char_level:tn  91.12±104.18  0.000000  921.0
        char_level:fp     1.45±3.55  0.000000   46.0
        char_level:fn     5.21±7.14  0.000000   64.0
 char_level:precision     0.93±0.18  0.021277    1.0
    char_level:recall     0.86±0.10  0.142857    1.0
        char_level:f1     0.87±0.15  0.041667    1.0
 word_level:precision     0.75±0.23  0.000000    1.0
    word_level:recall     0.67±0.23  0.000000    1.0
        word_level:f1     0.70±0.23  0.000000    1.0
```

**fix-newmm-longtext**

Colab : [Google colab](https://colab.research.google.com/drive/16z1pXRkSPeTNaQNcH8T6LFn_FOTg_00n)

```
============== Benchmark Result ==============
               metric      mean±std       min    max
        char_level:tp   25.72±27.76  1.000000  242.0
        char_level:tn  90.11±102.38  0.000000  915.0
        char_level:fp     2.46±4.72  0.000000   50.0
        char_level:fn     4.98±6.69  0.000000   59.0
 char_level:precision     0.91±0.18  0.021277    1.0
    char_level:recall     0.87±0.10  0.142857    1.0
        char_level:f1     0.87±0.15  0.041667    1.0
 word_level:precision     0.73±0.23  0.000000    1.0
    word_level:recall     0.67±0.23  0.000000    1.0
        word_level:f1     0.69±0.23  0.000000    1.0
```

File: pythainlp/tokenize/newmm.py [hunk 2]
```diff
@@ -96,12 +100,13 @@ def _onecut(text: str, custom_dict: Trie) -> Iterable[str]:
             heappush(q, i)
 
 
-def segment(text: str, custom_dict: Trie = None) -> List[str]:
+def segment(text: str, custom_dict: Trie = DEFAULT_DICT_TRIE) -> List[str]:
```
Comment (wannaphongcom) [inline pythainlp/tokenize/newmm.py]: เพิ่มพารามิเตอร์เข้าไปด้วยดีไหมครับ แบบถ้าเป็นประโยคยาว ๆ ให้กำหนด _TEXT_LIMIT เองครับ

File: pythainlp/tokenize/newmm.py [hunk 3]
```diff
@@ -21,10 +21,10 @@
 from .tcc import tcc_pos
 from .trie import Trie
 
-# ช่วยตัดพวกภาษาอังกฤษ เป็นต้น
+# To tokenizes English words, for example
```
Comment (heytitle) [inline pythainlp/tokenize/newmm.py]: `tokenize` without`s`?

File: pythainlp/tokenize/newmm.py [hunk 4]
```diff
@@ -110,4 +115,52 @@ def segment(text: str, custom_dict: Trie = None) -> List[str]:
...
+        while text_len >= (_TEXT_LIMIT + _TEXT_SCAN_RIGHT):
+            sample_start = _TEXT_LIMIT - _TEXT_SCAN_LEFT
+            sample_end = _TEXT_LIMIT + _TEXT_SCAN_RIGHT
+            sample = text[sample_start:sample_end]
+
+            # find possible break positions
+            cut_pos = sample_end
+
+            # try to break by space first
+            space_idx = sample.rfind(" ")
```
Comment (heytitle) [inline pythainlp/tokenize/newmm.py]: what happens if the space is in the middle of the name, e.g. ` ....ชื่อ นามสกุล.....`? Am I right that the algorithm probably tokenises wrongly in this case?

Comment (bact) [reply pythainlp/tokenize/newmm.py]: Yes. I will unfortunately breaks at the space.
This will also include case like "ปอน ๆ"

**Comment by bact** [comment]
Add `safe_mode` parameter, to enable/disable the preprocessing.
  - If `safe_mode` is `True`, do the preprocess (will be slower)
  - If `safe_mode` is `False` (default), don't do the preprocess (may face long wait for edge cases)

**Comment by wannaphongcom** [comment]
It's well. If this pull request is ready, You can merge.

File: pythainlp/tokenize/newmm.py (hunk 2 above)
Comment (bact) [reply pythainlp/tokenize/newmm.py]: ใช้เป็น safe_mode แทนละครับตอนนี้ เวิร์กไหม

**Pull request merged**